scala.wake

```diff
@@ -337,143 +346,142 @@ global def getScalaModuleTransIvyDeps module =
+  # Relative paths for running the job in buildDIr
+  def relMillBin = relative buildDir millBin.getPathName
+  def relMillHome = relative buildDir millHome
+  def relIvyCache = relative buildDir ivyCache
 
-# Compiles a ScalaModule, returns bloop config and generated jar file
-target memoizeBloopCompile name \ module =
-  def bin = bloopWake
   def transDeps = module.getScalaModuleTransDeps
-  def configTemplate = module.bloopTemplate
-  def configFilename = module.bloopConfigName
-  def jarDir = mkdir "{module.buildDir}/jar"
-  def jarName = "{jarDir.getPathName}/{name}.jar"
-  def inputs =
+  def buildFile = millBuildFile name module
+  def inputsResult =
     def sourceFiles = module.getScalaModuleAllSourceFiles ++ module.getScalaModuleResourceFiles
-    def depJars = map compileScalaModule transDeps
-    def configs = map bloopConfig transDeps
-    bin, configTemplate, configs ++ depJars ++ sourceFiles
-    | sortBy (_.getPathName <* _.getPathName) # Sorted to ensure hash is stable
-  # This is a hack to ensure that changes to visible files rerun the job
-  # We give directories as inputs to Bloop which then looks in those directories so new files m
-  def visibleHashHack =
-    def hash = (\x prim "hash") inputs
-    "VISIBLE_FILES_HASH_HACK={format hash}"
-  def bloopArgs =
-    def args = "compile", "-c", bloopDir.getPathName, module.getScalaModuleName, Nil
-    args | catWith " "
-  def cmd =
-    "python3", bin.getPathName, "-t", configTemplate.getPathName, "-c", configFilename , "-b", bloopInstall
-    , "--coursier-cache", ivyCache, "--bloop-args", bloopArgs, "--jar", jarName, Nil
-  def plan =
-    makePlan cmd inputs
-    | editPlanEnvironment ("JAVA_OPTS=-Xss8M", visibleHashHack, _)
-    | setPlanFnInputs (\_ inputs | map getPathName)
-    | setPlanLocalOnly True
-    | setPlanFnOutputs (\_ configFilename, jarName, Nil)
-    | setPlanResources resources
-    | runJob
-  def err paths = Fail (makeError "Expected exactly one json and one jar, got {format paths}")
-  def result = match plan.getJobOutputs
-    json, jar, Nil = Pass (Pair json jar)
-    # If the job fails, it returns 1 BadPath
-    one, Nil       = one.getPathError | omap Fail | getOrElse (err (one, Nil))
-    l              = err l
-  # Compose possible errors
-  module.scalaModuleToResult
-  | rmapPass (\_ result)
-def bloopCompile module = memoizeBloopCompile module.getScalaModuleName module
-
-def bloopConfig module = match module.bloopCompile
-  Pass (Pair f _) = f
-  Fail err        = makeBadPath err
-
-global def compileScalaModule module = match module.bloopCompile
-  Pass (Pair _ s) = s
-  Fail err        = makeBadPath err
+    def depJarsResult = transDeps | map compileScalaModule | map getPathResult | findFail
+    def addOtherInputs depJars =
+      millBin, buildFile, depJars ++ sourceFiles ++ (ivyCacheDeps Unit)
+      | sortBy (_.getPathName <* _.getPathName) # Sorted to ensure hash is stable
+    depJarsResult
+    | rmap addOtherInputs
```

Comment:
Many of these `rmap`s are hard to avoid because your actual return type for this function is `Path`, and the `require` construct is not easy to use for destructuring the error out of a Fail in the `else` case. For example, you can't do:

```wake
def maybeFoo = Fail "error".makeError
# This is pseudo-Scala code on the right half
require Pass foo = maybeFoo else { case Fail(e) => e.makeBadPath }
```

Even if you know that in the `else` case the `maybeFoo` is a `Fail`, you can't access the thing it has wrapped underneath.

I think what would probably simply things is splitting this into two functions, an outer lair that returns `Path` and an inner layer that returns `Result Path Error`. In the inner layer, you can always just return the `Fail` through the use of `require`, and in the outer layer, it's just one `match` expression.

Then you can do things like:

```wake
require Pass depJars = transDeps | map compileScalaModule | map getPathResult | findFail
def inputsResult = addOtherInputs depJars
def finalReturnValue =
  pretty
  | much
  | all
  | the
  | other stuff
Pass finalReturnValue
```